lib/bolt/pal/yaml_plan/step.rb

```diff
@@ -0,0 +1,190 @@
+# frozen_string_literal: true
+
+require 'bolt/util'
+
+module Bolt
+  class PAL
+    class YamlPlan
+      class Step
+        attr_reader :name, :type, :body, :target
+
+        COMMON_STEP_KEYS = %w[name description target].freeze
+        STEP_KEYS = {
+          'command' => {
+            'allowed_keys' => Set['command'].merge(COMMON_STEP_KEYS),
+            'required_keys' => Set['target']
+          },
+          'script' => {
+            'allowed_keys' => Set['script', 'parameters', 'arguments'].merge(COMMON_STEP_KEYS),
+            'required_keys' => Set['target']
+          },
+          'task' => {
+            'allowed_keys' => Set['task', 'parameters'].merge(COMMON_STEP_KEYS),
+            'required_keys' => Set['target']
+          },
+          'plan' => {
+            'allowed_keys' => Set['plan', 'parameters'].merge(COMMON_STEP_KEYS),
+            'required_keys' => Set.new
+          },
+          'source' => {
+            'allowed_keys' => Set['source', 'destination'].merge(COMMON_STEP_KEYS),
+            'required_keys' => Set['target', 'source', 'destination']
+          },
+          'destination' => {
+            'allowed_keys' => Set['source', 'destination'].merge(COMMON_STEP_KEYS),
+            'required_keys' => Set['target', 'source', 'destination']
+          },
+          'eval' => {
+            'allowed_keys' => Set['eval', 'name', 'description'],
+            'required_keys' => Set.new
+          }
+        }.freeze
+
+        def initialize(step_body, step_number)
+          @body = step_body
+          @name = @body['name']
+          # For error messages
+          @step_number = step_number
+          validate_step
+
+          @type = STEP_KEYS.keys.find { |key| @body.key?(key) }
+          @target = @body['target']
+        end
+
+        def transpile(plan_path)
+          result = String.new("  ")
+          result << "$#{@name} = " if @name
+
+          description = body.fetch('description', nil)
+          parameters = body.fetch('parameters', {})
+          if @type == 'script' && body.key?('arguments')
+            parameters['arguments'] = body['arguments']
+          end
+
+          case @type
+          when 'command', 'task', 'script', 'plan'
+            result << "run_#{@type}(#{Bolt::Util.to_code(body[@type])}"
+            result << ", #{Bolt::Util.to_code(@target)}" if @target
+            result << ", #{Bolt::Util.to_code(description)}" if description
+            result << ", #{Bolt::Util.to_code(parameters)}" unless parameters.empty?
```

Comment:
I think we are missing a validation here that parameters is in fact an array. If it is not calling `.empty?` will stack trace.

Reply:
Do you have a plan where this happens? Parameters should get set to an empty hash by default, and we have a check in the yaml plan that it's a hash...

Reply:
I think somehow we are just not validating that parameters is a hash. 
```
parameters:
  nodes:
    type: TargetSpec
steps:
  - name: run_task
    task: sample::echo
    description: "foo"
    target: localhost
    parameters: foo
```
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt plan convert Boltdir/site/yaml_plans/plans/test_convert.yaml 
bundler: failed to load command: bolt (/home/cas/working_dir/bolt/.bundle/ruby/2.5.0/bin/bolt)
NoMethodError: undefined method `empty?' for #<Bolt::PAL::YamlPlan::BareString:0x000055fe834b08e8>
  /home/cas/working_dir/bolt/lib/bolt/pal/yaml_plan/step.rb:69:in `transpile'
  /home/cas/working_dir/bolt/lib/bolt/pal/yaml_plan/transpiler.rb:41:in `block in transpile'
  /home/cas/working_dir/bolt/lib/bolt/pal/yaml_plan/transpiler.rb:39:in `each'
  /home/cas/working_dir/bolt/lib/bolt/pal/yaml_plan/transpiler.rb:39:in `transpile'
  /home/cas/working_dir/bolt/lib/bolt/pal.rb:303:in `convert_plan'
  /home/cas/working_dir/bolt/lib/bolt/cli.rb:465:in `convert_plan'
  /home/cas/working_dir/bolt/lib/bolt/cli.rb:283:in `execute'
  /home/cas/working_dir/bolt/exe/bolt:10:in `<top (required)>'
  /home/cas/working_dir/bolt/.bundle/ruby/2.5.0/bin/bolt:23:in `load'
  /home/cas/working_dir/bolt/.bundle/ruby/2.5.0/bin/bolt:23:in `<top (required)>'
```

Reply:
Oh, I see, we're validating *plan* parameters not step parameters.